src/compute/src/logging/reachability.rs

```diff
@@ -155,10 +91,115 @@ pub fn construct<A: Allocate>(
+
+                let mut buffer = Vec::new();
+                flatten.build(move |_capability| {
+                    move |_frontiers| {
+                        let updates = updates_out.activate();
+                        let mut updates_session = ConsolidateBuffer::new(updates, 0);
+
+                        input.for_each(|cap, data| {
+                            data.swap(&mut buffer);
+
+                            for (time, worker, (addr, massaged)) in buffer.drain(..) {
+                                let time_ms = (((time.as_millis() as Timestamp / granularity_ms)
+                                    + 1)
+                                    * granularity_ms)
+                                    as Timestamp;
+                                for (source, port, update_type, ts, diff) in massaged {
+                                    updates_session.give(
+                                        &cap,
+                                        (
+                                            (update_type, addr.clone(), source, port, worker, ts),
+                                            time_ms,
+                                            diff,
+                                        ),
+                                    );
+                                }
+                            }
+                        });
+                    }
+                });
+
+                let common = updates
+                    .as_collection()
+                    .arrange_core::<_, RowSpine<_, _, _, _>>(
+                        Exchange::new(|(((_, _, _, _, w, _), ()), _, _)| *w as u64),
+                        "PreArrange Timely reachability",
+                    );
+
+                if is_active {
+                    // Use common to create an arrangement.
+                    let mut row_buf = Row::default();
+                    let updates = common.as_collection(
+                        move |(update_type, addr, source, port, worker, ts), _| {
+                            let row_arena = RowArena::default();
+                            let update_type = if *update_type { "source" } else { "target" };
+                            row_buf.packer().push_list(
+                                addr.iter()
+                                    .chain_one(&source)
+                                    .map(|id| Datum::Int64(*id as i64)),
+                            );
+                            let datums = &[
+                                row_arena.push_unary_row(row_buf.clone()),
+                                Datum::Int64(*port as i64),
+                                Datum::Int64(*worker as i64),
+                                Datum::String(&update_type),
+                                Datum::from(ts.and_then(|ts| i64::try_from(ts).ok())),
+                            ];
+                            row_buf.packer().extend(key.iter().map(|k| datums[*k]));
```

Comment:
Please move the non-variant dependent code outside of the loop. We currently only have a single log source here, but if this changes in the future, we want to make sure that each dataflow only gets constructed once.